### ISSUE TITLE
[cloud_firestore] Writebatch serverTimestamp crashing offline

### ISSUE THREAD

**Issue by Minituff** [FirebaseExtended/flutterfire#1322]
I am working an app that allows the user to send messages. In order to ensure timestamp consistency I am using the serverTimestamp feature. 

This is the stream I am using to load new messages as they are sent:
```dart
Firestore.instance
        .collection('$chatRoomPath/messages')
        .where('sent_at', isGreaterThan: lastMsgSentAt)
        .snapshots();
```

Here is a code snippet I am using for when a new message is created:
```dart
final WriteBatch batch = _firestore.batch();

final DocumentReference newMsgDoc = _firestore.document('$chatRoomPath/messages/$newDocID');

// Update another doc in batch
batch.setData(someOtherDoc, {
      'foo': bar,
});

// Add a new msgDoc
batch.setData(newMsgDoc, {
      'text': text,
      'sent_at': FieldValue.serverTimestamp(),
});

//Commit both writes at once
return batch.commit();
```

It should be noted that this **DOES** work when I dont use a writebatch

This works great online, but when the user is offline, I think the app is crashing since it is trying to query a fieldvalue that doesn't yet exist. 

I am using   **cloud_firestore: ^0.12.9+5**

Here is a crash I get when I attempt to send a message in offline mode:
```java
E/AndroidRuntime(19173): Process: com.xxx.xxx, PID: 19173
E/AndroidRuntime(19173): java.lang.RuntimeException: Internal error in Firestore (19.0.0).
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.util.AsyncQueue.lambda$panic$5(com.google.firebase:firebase-firestore@@19.0.0:379)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.util.AsyncQueue$$Lambda$5.run(Unknown Source:2)
E/AndroidRuntime(19173): 	at android.os.Handler.handleCallback(Handler.java:873)
E/AndroidRuntime(19173): 	at android.os.Handler.dispatchMessage(Handler.java:99)
E/AndroidRuntime(19173): 	at android.os.Looper.loop(Looper.java:193)
E/AndroidRuntime(19173): 	at android.app.ActivityThread.main(ActivityThread.java:6863)
E/AndroidRuntime(19173): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(19173): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:537)
E/AndroidRuntime(19173): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
E/AndroidRuntime(19173): Caused by: java.lang.AssertionError: INTERNAL ASSERTION FAILED: Encoded field value should not be null.
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.util.Assert.fail(com.google.firebase:firebase-firestore@@19.0.0:46)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.util.Assert.hardAssert(com.google.firebase:firebase-firestore@@19.0.0:31)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.remote.RemoteSerializer.encodeValue(com.google.firebase:firebase-firestore@@19.0.0:263)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.remote.RemoteSerializer.encodeDocument(com.google.firebase:firebase-firestore@@19.0.0:389)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.remote.RemoteSerializer.encodeMutation(com.google.firebase:firebase-firestore@@19.0.0:436)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.local.LocalSerializer.encodeMutationBatch(com.google.firebase:firebase-firestore@@19.0.0:169)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.local.SQLiteMutationQueue.addMutationBatch(com.google.firebase:firebase-firestore@@19.0.0:188)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.local.LocalStore.lambda$writeLocally$1(com.google.firebase:firebase-firestore@@19.0.0:237)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.local.LocalStore$$Lambda$2.get(Unknown Source:8)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.local.SQLitePersistence.runTransaction(com.google.firebase:firebase-firestore@@19.0.0:195)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.local.LocalStore.writeLocally(com.google.firebase:firebase-firestore@@19.0.0:201)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.core.SyncEngine.writeMutations(com.google.firebase:firebase-firestore@@19.0.0:227)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.core.FirestoreClient.lambda$write$11(com.google.firebase:firebase-firestore@@19.0.0:200)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.core.FirestoreClient$$Lambda$11.run(Unknown Source:6)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$4(com.google.firebase:firebase-firestore@@19.0.0:311)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.util.AsyncQueue$$Lambda$4.call(Unknown Source:2)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$3(com.google.firebase:firebase-firestore@@19.0.0:287)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.util.AsyncQueue$$Lambda$3.run(Unknown Source:4)
E/AndroidRuntime(19173): 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:458)
E/AndroidRuntime(19173): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E/AndroidRuntime(19173): 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
E/AndroidRuntime(19173): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
E/AndroidRuntime(19173): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
E/AndroidRuntime(19173): 	at com.google.firebase.firestore.util.AsyncQueue$DelayedStartFactory.run(com.google.firebase:firebase-firestore@@19.0.0:205)
E/AndroidRuntime(19173): 	at java.lang.Thread.run(Thread.java:764)
E/libc    (19173): Access denied finding property "vendor.debug.egl.swapinterval"
W/OPDiagnose(19173): getService:OPDiagnoseService NULL
D/OSTracker(19173): OS Event: crash
D/AbstractTracker(19173): Event success
I/Process (19173): Sending signal. PID: 19173 SIG: 9
Lost connection to device.
Exited (sigterm)
```

In the other Firebase SDKs they have a date estimation feature which looks like it would solve this problem, but as far as I know this feature is not available yet for the Flutter SDK.

```java
DocumentSnapshot.ServerTimestampBehavior behavior = ESTIMATE;
Date date = snapshot.getDate("OrderDate", behavior);
```

Thank you for your help.

*Edit:* Added writebatch, I previously left it out for brevity, but it seems like this is actually the main source of the problem.

This seems to be a related issue: https://github.com/FirebaseExtended/flutterfire/issues/558

**Comment by yvzoncu** [comment]
i having the same issue while trying to create a new document and setData in it.

**Comment by JaiMistry** [comment]
I am facing this exact same issue as well on the latest version of Firestore.

**Comment by iapicca** [comment]
Hi @Minituff 
could you please provide your `flutter doctor -v`
Thank you

**Comment by Minituff** [comment]
Hi @iapicca,

Here's my output:
```PS
PS C:\Users\James-PC\Projects\Project1> flutter doctor -v
[√] Flutter (Channel stable, v1.9.1+hotfix.6, on Microsoft Windows [Version 10.0.17763.805], locale en-US)
    • Flutter version 1.9.1+hotfix.6 at C:\Users\James-PC\flutter
    • Framework revision 68587a0916 (8 weeks ago), 2019-09-13 19:46:58 -0700
    • Engine revision b863200c37
    • Dart version 2.5.0

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at C:\Users\James-PC\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)
    • All Android licenses accepted.

[√] Android Studio (version 3.5)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 41.0.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)

[√] IntelliJ IDEA Ultimate Edition (version 2019.2)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA 2019.2.4
    • Flutter plugin version 40.2.4
    • Dart plugin version 192.7402

[√] VS Code (version 1.39.2)
    • VS Code at C:\Users\James-PC\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.6.0

[√] Connected device (1 available)
    • Android SDK built for x86 64 • emulator-5554 • android-x64 • Android 6.0 (API 23) (emulator)

• No issues found!
```

**Comment by Salakar** [comment]
Hey all, as part of our on-going work for #2582, this has been resolved in our Firebase Firestore rework (#2913) - which has been merged into master. We'll look at publishing some prereleases in the next few days. Thank you
